src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs

```diff
@@ -259,7 +234,78 @@ IEnumerator IEnumerable.GetEnumerator()
             }
         }
 
-        private class ConfigurationBuilderProperties : IDictionary<string, object>
+        private sealed class ProviderManager : IDisposable
+        {
+            private readonly object _replaceProvidersLock = new object();
+            private ProvidersReference _refCountedProviders = new(new List<IConfigurationProvider>());
+
+            public IEnumerable<IConfigurationProvider> Providers => _refCountedProviders.Providers;
+
+            public ProvidersReference GetReference()
+            {
+                // Lock to ensure oldRefCountedProviders.Dispose() in ReplaceProviders() doesn't decrement ref count to zero
+                // before calling _refCountedProviders.AddRef().
+                lock (_replaceProvidersLock)
+                {
+                    _refCountedProviders.AddRef();
+                    return _refCountedProviders;
+                }
+            }
+
+            // Providers should never be concurrently modified. Reading during modification is allowed.
+            public void ReplaceProviders(List<IConfigurationProvider> providers)
+            {
+                ProvidersReference oldRefCountedProviders = _refCountedProviders;
+
+                lock (_replaceProvidersLock)
+                {
+                    _refCountedProviders = new ProvidersReference(providers);
+                }
+
+                oldRefCountedProviders.Dispose();
+            }
+
+            public void AddProvider(IConfigurationProvider provider)
+            {
+                // Maintain existing references, but replace list with copy containing new item.
+                _refCountedProviders.Providers = new List<IConfigurationProvider>(_refCountedProviders.Providers)
```

Comment:
Would it make sense to initialize this list with `_refCountedProviders.Providers.Count + 1` capacity? That way when you add the new provider, if we are at a bound (like going from 4 to 5), we don't immediately grow the backing array.

Reply:
That's a pretty cool idea. I'm not sure it's worth the effort though. There's a lot of other work involved in adding a configuration source.